watcher/src/bin/avr_export.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! A utility for exporting the history of Attestation Verification
+//! Reports generated by MobileCoin consensus nodes
+
+use clap::Parser;
+use mc_blockchain_validators::metadata::avr::{AvrConfig, AvrConfigRecord};
+use mc_common::{
+    logger::{create_app_logger, o},
+    ResponderId,
+};
+use mc_crypto_keys::Ed25519Public;
+use mc_watcher::{error::WatcherDBError, watcher_db::WatcherDB};
+use std::{fs, path::PathBuf};
+use url::Url;
+
+/// Command line configuration.
+#[derive(Parser)]
+#[clap(
+    name = "mc-watcher-avr-export",
+    about = "A utility for exporting the history of MobileCoin consensus enclave AVRs"
+)]
+pub struct Config {
+    /// Path to watcher db (lmdb).
+    #[clap(
+        long,
+        default_value = "watcher-db",
+        parse(from_os_str),
+        env = "MC_WATCHER_DB"
+    )]
+    pub watcher_db: PathBuf,
+
+    /// Path for the avr-history.toml & avr-history.json bootstrap files to be
+    /// written.
+    #[clap(
+        long,
+        default_value = "",
+        parse(from_os_str),
+        env = "MC_AVR_HISTORY_PATH"
+    )]
+    pub avr_history: PathBuf,
+}
+
+fn main() {
+    let (logger, _global_logger_guard) = create_app_logger(o!());
+
+    let mut config = Config::parse();
+    config.avr_history.set_file_name("avr_history");
+    let watcher_db =
+        WatcherDB::open_ro(&config.watcher_db, logger).expect("Failed opening watcher db");
+    let mut avr_reports = AvrConfig::default();
+
+    // Get all of the latest synced blocks
+    let last_synced_blocks = watcher_db.last_synced_blocks().unwrap();
+
+    // Find where the signing keys changed and fetch the AVRs for those keys
+    for (tx_src_url, max_block_index) in last_synced_blocks.iter() {
```

Comment:
> @eranrund commented
> I believe this skips writing an entry with no end index for the last signer

Reply:
That might be something we want to fill in manually - although in practice we could automate it here, we have to be REALLY sure that the watcherdb we're bootstrapping from is the absolutely correct current record. It'd be easy to create a toml file here that's outdated. This database was dump ops had to make manually and as of yet there's not automated process.

My intuition for fully automating it is to:

Make an attest call to the consensus nodes for their current AVRs every time this bootstrap is run, and maybe double check it every time we generate one before publishing it
Ensure this bootstrap script is maybe run on a watcher node whose database has the most complete records.
It's a painful problem right now, but when metadata on chain arrives, this toml file will be purely historical, and this issue of "how do we release a toml file with the most current avrs" ceases to be necessary. In the meantime, we do need to decide on it.

I think I'd probably advocate an approach where we don't rely on the TOML file to be authoritative about the present, have it be strictly historical, and start recommending clients do metadata validation only when it begins appearing on chain. Opinions may differ on this however.